cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java

```diff
@@ -0,0 +1,381 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(ExecutionSparkContext.class);
+
+  // Cache of Dataset instances that are created through this context
+  // for closing all datasets when closing this context. This list does not includes datasets
+  // created through the readFromDataset and writeToDataset methods.
+  private final Map<String, Dataset> datasets;
+  private final Configuration hConf;
+  private final Transaction transaction;
+  private final StreamAdmin streamAdmin;
+  private final DatasetInstantiator datasetInstantiator;
+  private boolean stopped;
+  private ApacheSparkContext apacheSparkContext;
+
+  public ExecutionSparkContext(SparkSpecification specification, Id.Program programId, RunId runId,
+                               ClassLoader programClassLoader, long logicalStartTime,
+                               Map<String, String> runtimeArguments,
+                               Transaction transaction, DatasetFramework datasetFramework,
+                               DiscoveryServiceClient discoveryServiceClient,
+                               MetricsCollectionService metricsCollectionService,
+                               Configuration hConf, StreamAdmin streamAdmin) {
+    this(specification, programId, runId, programClassLoader, logicalStartTime, runtimeArguments,
+         transaction, datasetFramework, discoveryServiceClient,
+         createMetricsContext(metricsCollectionService, programId, runId),
+         createLoggingContext(programId, runId), hConf, streamAdmin);
+  }
+
+  public ExecutionSparkContext(SparkSpecification specification, Id.Program programId, RunId runId,
+                                ClassLoader programClassLoader, long logicalStartTime,
+                                Map<String, String> runtimeArguments,
+                                Transaction transaction, DatasetFramework datasetFramework,
+                                DiscoveryServiceClient discoveryServiceClient, MetricsContext metricsContext,
+                                LoggingContext loggingContext, Configuration hConf, StreamAdmin streamAdmin) {
+    super(specification, programId, runId, programClassLoader, logicalStartTime,
+          runtimeArguments, discoveryServiceClient, metricsContext, loggingContext);
+
+    this.datasets = new HashMap<>();
+    this.hConf = hConf;
+    this.transaction = transaction;
+    this.streamAdmin = streamAdmin;
+    this.datasetInstantiator = new DatasetInstantiator(programId.getNamespace(), datasetFramework,
+                                                       programClassLoader, getOwners(), getMetricsContext());
+  }
+
+  @Override
+  public <T> T readFromDataset(String datasetName, Class<?> kClass, Class<?> vClass) {
+    Map<String, String> datasetArgs = RuntimeArguments.extractScope(Scope.DATASET, datasetName, getRuntimeArguments());
+
+    // Verify the dataset exists. If it doesn't exist, exception will be thrown
```

Comment:
Actually this call is not needed. Not sure how it gets there, probably because I was testing something else.